File: Freedom/微信/通讯录/WXFriendSearchViewController.m
//  TLFriendSearchViewController.m
//  Freedom
//  Created by Super on 16/1/25.
#import "WXFriendSearchViewController.h"
#import "WXFriendsViewController.h"
@interface WXFriendSearchViewController ()
@property (nonatomic, strong) NSMutableArray *data;
@end
@implementation WXFriendSearchViewController
- (void)viewDidLoad{
    [super viewDidLoad];
    
    self.data = [[NSMutableArray alloc] init];;
    [self.tableView registerClass:[WXFriendCell class] forCellReuseIdentifier:@"FriendCell"];
    [self setAutomaticallyAdjustsScrollViewInsets:NO];
}
- (void) viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    self.tableView.frameY = HEIGHT_NAVBAR + NavY;
    self.tableView.frameHeight = APPH - self.tableView.frameY;
}
#pragma mark - UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return self.data.count;
}
- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section{
    return @"联系人";
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    WXFriendCell *cell = [tableView dequeueReusableCellWithIdentifier:@"FriendCell"];
    
    WXUser *user = [self.data objectAtIndex:indexPath.row];
    [cell setUser:user];
    [cell setTopLineStyle:(indexPath.row == 0 ? TLCellLineStyleFill : TLCellLineStyleNone)];
    [cell setBottomLineStyle:(indexPath.row == self.data.count - 1 ? TLCellLineStyleFill : TLCellLineStyleDefault)];
    return cell;
}
#pragma mark - Delegate -
//MARK: UITableViewDelegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return HEIGHT_FRIEND_CELL;
}
//MARK: UISearchResultsUpdating
- (void)updateSearchResultsForSearchController:(UISearchController *)searchController {
    NSString *searchText = [searchController.searchBar.text lowercaseString];
    [self.data removeAllObjects];
    for (WXUser *user in self.friendsData) {
        if ([user.remarkName containsString:searchText] || [user.username containsString:searchText] || [user.nikeName containsString:searchText] || [user.pinyin containsString:searchText] || [user.pinyinInitial containsString:searchText]) {
            [self.data addObject:user];
        }
    }
    [self.tableView reloadData];
}
@end

File: Freedom/微信/通讯录/WXFriendSearchViewController.swift
//
//  WXFriendSearchViewController.swift
//  Freedom

import Foundation
class WXFriendSearchViewController: BaseTableViewController, UISearchResultsUpdating {
    var friendsData: [WXUser] = []
    var data: [WXUser] = []
    override func viewDidLoad() {
        super.viewDidLoad()
        tableView.register(WXFriendCell.self, forCellReuseIdentifier: WXFriendCell.identifier)
        automaticallyAdjustsScrollViewInsets = false
    }
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        tableView.frame = CGRect(x: tableView.frame.origin.x, y: TopHeight + 20, width: tableView.frame.size.width, height: APPH - tableView.frame.origin.y)
    }
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return data.count
    }
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String {
        return "联系人"
    }
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueCell(WXFriendCell.self)
        let user = data[indexPath.row]
        cell.viewModel.onNext(user)
        return cell
    }
    override func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 54
    }
    func updateSearchResults(for searchController: UISearchController) {
        let searchText = searchController.searchBar.text?.lowercased() ?? ""
        data.removeAll()
        for user: WXUser in friendsData {
            if user.remarkName.contains(searchText) || user.username.contains(searchText) || user.nikeName.contains(searchText) || user.nikeName.pinyin().contains(searchText) || user.nikeName.firstLetter().contains(searchText) {
                data.append(user)
            }
        }
        tableView.reloadData()
    }
}
